apps/api/src/server/applications/subscriptions/subscriptions.service.js

```diff
@@ -15,11 +17,11 @@ import { EventType } from '@pins/event-client';
  * @returns {Promise<{id: number, created: boolean}>}
  */
 export async function createOrUpdateSubscription(request) {
-	const subscription = prepareInput(request);
+	const subscription = await prepareInput(request);
 
 	const existing = await subscriptionRepository.findUnique(
-		subscription.caseReference,
-		subscription.emailAddress
+		request.caseReference,
+		request.emailAddress
```

Comment:
I had a look into doing this, and we'd need to use an [interactive transaction](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#interactive-transactions) since we need to run `.update()` if it already exists and `.create()` if not. The problem is inside a transaction it looks like you can only use the Prisma methods directly, we can't use the methods we've defined ourselves in subscription.respository.js, so it would be quite a verbose function.

Do you think it's worthwhile, since we're just running a query (doesn't mutate) followed by a single mutation, so it shouldn't be unsafe the way it currently is.

Reply:
It's not unsafe per-se, just not atomic. In the unlikely event another subscription is created between find + create, we might end up with two or an error.

There's an example here where we make sure a project update status change is allowed in a transaction: https://github.com/Planning-Inspectorate/back-office/blob/main/apps/api/src/server/repositories/project-update.respository.js#L126

Let's make this a separate ticket perhaps - there are probably other places we should be doing this.

Reply:
Yeah agreed. I've made a ticket for this https://pins-ds.atlassian.net/browse/BOAS-1421